tests.trivial-builders.references: refactor and move expressions into trivial-builders/test/references

ShamrockLee opened the pull request:
###### Description of changes
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Make a `testScriptBin` that contains the substituted test script.
* Add an `installCheckPhase` to check the result script with ShellCheck.
* Passthru as `references.testScriptBin` to run the (substituted) test script directly (without VM).
* Drop the logic in build script that detects if it is run in the Nix sandbox.
* Inline sample application; drop invoke-*.nix.

Format expressions.
* Format with `nixpkgs-fmt`.
* Use multi-line style of set patterns.

Call the `samples` with `callPackage`.
* Rename `sample` to `samples`.
* Take individual packages / build helpers directly from the set pattern.
* Define `cleanSamples` to filter out overriders such as `<pkg>.override` added by `callPackage`.

Passthru samples and invocation results for easier debugging.
* Passthru `samples`, `references` and `directReferences`

Move all the related files into directory `trivial-builders/test/reference`
* Make the `trivial-builders/test` directory cleaner, and `samples.nix` not ambiguous.
* Call it with `callPackage ./references { }` instead of `callPackaeg ./references.nix` in `trivial-builders/test/default.nix`.
* Pass it into `tests.trivial-builders.writeStringReferencesToFile`

Cc: @roberth

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. --> 

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ShamrockLee committed:
trivial-builders/test/references-test.sh: lint with ShellCheck

ShamrockLee committed:
tests.trivial-builders.references: refactor expression

Define package `testScriptBin` that contains the substituted test script.
* Add an `installCheckPhase` to check the result script with ShellCheck.
* Passthru as `references.testScriptBin` to run the
  (substituted) test script directly (without VM).
* Drop the logic in build script that detects if
  it is run in the Nix sandbox.
* Inline sample application; drop invoke-*.nix.

Format expressions.
* Format with `nixpkgs-fmt`.
* Use multi-line style of set patterns.

Call the samples with `callPackage`.
* Rename `sample` -> `samples`.
* Take individual packages / build helpers directly from the
  set pattern.
* Define `cleanSamples` to filter out overriders such as `<pkg>.override`.
  added by `callPackage`.

Passthru samples and invocation results for easier debugging.
* Passthru samples, references, directReferences
* Provide tests.trivial-builders.writeStringReferencesToFile with such
  samples argument.

ShamrockLee commented:
Rebase onto the master branch.

ShamrockLee committed:
tests.trivial-buildrs.references: move expression files into trivial-builders/test/references/

ShamrockLee committed:
tests.trivial-builders.references: add ShamrockLee as a maintainer

roberth commented on pkgs/build-support/trivial-builders/test/default.nix:
```diff
@@ -14,21 +14,24 @@
     if stdenv.hostPlatform.isLinux
-    then callPackage ./references.nix {}
+    then references
     else null;
```
```suggestion
    else {};
```

It seems that this would please ofborg:
https://github.com/NixOS/nixpkgs/pull/204462/checks?check_run_id=19453355734

roberth reviewed (approved):
This looks alright!

ShamrockLee committed:
tests.trivial-builders.references: specify as empty set instead of null on non-Linux

Please Nix CI (OfBorg) with empty set instead of null on non-linux platforms,
where NixOS tests are not supported.

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

roberth commented:
It works. Thanks!
